### ISSUE TITLE
Replace Farsi/Urdu characters with correct Arabic characters in annotated Arabic text

### ISSUE THREAD

**Issue by zeerakahmed** [zeerakahmed/makhzan#16]
In text identified as Arabic, make the following replacements:

1. ہ (U+06C1 ARABIC LETTER HEH GOAL) → (U+0647 ARABIC LETTER HEH)
2. ی (U+06CC ARABIC LETTER FARSI YEH) → ي (U+064A ARABIC LETTER YEH)

Identified in [discussion for another issue by @saadatm](https://github.com/zeerakahmed/makhzan/issues/10).

**Comment by saadatm** [comment]
Another replacement:

ک (U+06A9 ARABIC LETTER KEHEH) → ك (U+0643 ARABIC LETTER KAF)

> ی (U+06CC ARABIC LETTER FARSI YEH) → ي (U+064A ARABIC LETTER YEH)

There's also ى (U+0649 ARABIC LETTER ALEF MAKSURA), which is used in Arabic at the end of some words like أخرى,‎ عيسى, etc.

**Comment by zeerakahmed** [comment]
Good call on the Keheh to Kaf conversion.

For Yeh and Alef Maksura, I'm assuming that we can say for certain that Alef Maksura will only occur at the end of Arabic words, but I'm not certain we can say that Yeh will not occur at the end of words. Is that correct? Is there a good heuristic to determine whether the Farsi Yeh should be replaced with a Yeh or an Alef Maksura?

**Comment by saadatm** [comment]
Yes, I think that's correct, but no idea about any such heuristic. Let me request @khaledhosny to give us a few pointers here.

**Comment by khaledhosny** [comment]
I’m not sure about the source of the text or the reason of making these replacements, but I don’t think U+06CC is used in place of U+0649 since the later is always pronounced alif (the name farsi yeh is a misnomer, there are several Arabic language orthographies where final/isolated yeh is dotless).

**Comment by khaledhosny** [comment]
U+0649 can occur in the initial/medial position, but not i modern Arabic orthography, e.g. دعوىهم (instead of دعواهم) and بنىها (instead of بناها).

**Comment by saadatm** [comment]
> 
> 
> I’m not sure about the source of the text or the reason of making these replacements, but I don’t think U+06CC is used in place of U+0649 since the later is always pronounced alif (the name farsi yeh is a misnomer, there are several Arabic language orthographies where final/isolated yeh is dotless).

The background is: This project/repo is for an Urdu text corpus which may include some text in other languages too (Arabic, English, etc. for quotes or technical terms). Such non-Urdu text is being annotated in the corpus with an `<annotation>` tag and a respective `lang` attribute.

Now, the Arabic text currently in the corpus erroneously uses 06CC where either 064A or 0649 should be used. This is because out of all these _yeh_ characters, only 06CC is used in Urdu, and thus only 06CC is found in Urdu keyboards too. This has the side effect that when Urdu authors are quoting Arabic in their work, they make the mistake of using 06CC instead of 064A/0649.

We want to correct such mistakes in the corpus, but just replacing 06CC with 064A may be incorrect for words where 0649 should be used instead. So we are wondering if there is a heuristic to determine which of 064A and 0649 is appropriate in a given Arabic word.

**Comment by khaledhosny** [comment]
I see. In this case, I’m not aware of any reliable heuristic other than assuming all non-final U+06CC to be U+064A, but the reverse i snot true (both U+064A and U+0649 appear in final position, and even on otherwise identical words like على\علي).

**Comment by zeerakahmed** [comment]
Thanks @khaledhosny!

The general rule we've been following in Makhzan is to fix errors in the source text when we are confident we know the problem and the solution deterministically. Where unsure, we let the source text remain as is. So in this case I think our options would be to:

1. Leave all Arabic text as is
2. Correct all initial and medial use of U+06CC in Arabic text by replacing it with U+064A.

Thoughts?

**Comment by zeerakahmed** [comment]
By the way, @khaledhosny, I've used your work with multiple projects (Aref Ruqaa in particular). Thank you for it!

**Comment by khaledhosny** [comment]
I’d go with option 1, option 2 should be good for modern orthography but I’d be hesitant to do such a change without reviewing the text.

**Comment by zeerakahmed** [comment]
Thanks! We will only plan to make the following changes:

1. ہ (U+06C1 ARABIC LETTER HEH GOAL) → (U+0647 ARABIC LETTER HEH)
2. ک (U+06A9 ARABIC LETTER KEHEH) → ك (U+0643 ARABIC LETTER KAF)

**Comment by saadatm** [comment]
Forgot about this one:

ۃ (U+06C3 ARABIC LETTER TEH MARBUTA GOAL) → ة (U+0629 ARABIC LETTER TEH MARBUTA)
